clang/lib/Headers/hlsl/hlsl_detail.h

```diff
@@ -33,6 +41,23 @@ constexpr enable_if_t<sizeof(U) == sizeof(T), U> bit_cast(T F) {
   return __builtin_bit_cast(U, F);
 }
 
+template <typename T>
+constexpr enable_if_t<is_same<float, T>::value || is_same<half, T>::value, T>
+length_impl(T X) {
+  return __builtin_elementwise_abs(X);
+}
+
+template <typename T, int N>
+enable_if_t<is_same<float, T>::value || is_same<half, T>::value, T>
+length_vec_impl(vector<T, N> X) {
+  vector<T, N> XSquared = X * X;
+  T XSquaredSum = XSquared[0];
+  [unroll]
+  for (int i = 1; i < N; ++i)
```

Comment:
Because of the loop this can't be a constexpr. Made this change because its unlikely we will be able to use the vec reduce add builtin.

Reply:
This is because HLSL is only C++11 and not C++14, right? We could always implement this with template recursion if we really want it to be constexpr, but in all likelihood waiting until we bring HLSL up to C++14 eventually before we make this constexpr is fine.

Reply:
The way we have done HLSL language modes in `clang/include/clang/Basic/LangStandards.def` makes it seem like 202x and older language modes will always be C++11? Is that the wrong way of thinking about it? Will there be a day we change all language modes to C++14?